readthedocs/builds/constants.py

```diff
@@ -5,13 +5,15 @@
 
 
 BUILD_STATE_TRIGGERED = 'triggered'
+BUILD_STATE_PULLING_CACHE = 'pulling-cache'
 BUILD_STATE_CLONING = 'cloning'
 BUILD_STATE_INSTALLING = 'installing'
 BUILD_STATE_BUILDING = 'building'
 BUILD_STATE_FINISHED = 'finished'
 
 BUILD_STATE = (
     (BUILD_STATE_TRIGGERED, _('Triggered')),
+    (BUILD_STATE_PULLING_CACHE, _('Pulling cache')),
```

Comment:
It would be good to call this something that works for both pushing & pulling, perhaps `Updating cache`.

Reply:
If we want this, we will need to change how we handle the FINISHED status.

Currently, the build is first marked as FINISHED and then the push is performed. The FINISHED status is sent when we leave the context manager of the environment (`__exit__`) and the cache is pushed after that _only_ if the build was successful.

Taking a look at the code, it seems that our options are:

1. update the state of the build to FINISHED _outside_ the environment `__exit__` method
1. use `update_on_success=False` when initializing the environment and update the build status to FINISHED manually _after_ pushing the cache.
1. push cache always (not only on successful builds)

I think that 2) is our better option since it uses the API we already have in the code to avoid updating the status automatically.

If we are going in this direction, why not to add an specific status for each operation? Pulling/Pushing or Downloading/Uploading or similar.

Reply:
I went ahead and implemented 2).

It shows "Pushing cache" if the project has the feature and the build is marked as finished outside the `__exit__` method. With these changes, the total build time will count pulling/pushing the cache as well, so it will be more realistic.